src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractContinuousIntegrationService.java

```diff
@@ -0,0 +1,132 @@
+package de.tum.in.www1.artemis.service.connectors;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Optional;
+
+import javax.validation.constraints.NotNull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.tum.in.www1.artemis.domain.ProgrammingSubmission;
+import de.tum.in.www1.artemis.domain.Result;
+import de.tum.in.www1.artemis.domain.enumeration.AssessmentType;
+import de.tum.in.www1.artemis.domain.enumeration.SubmissionType;
+import de.tum.in.www1.artemis.domain.participation.Participation;
+import de.tum.in.www1.artemis.domain.participation.ProgrammingExerciseParticipation;
+import de.tum.in.www1.artemis.repository.ProgrammingSubmissionRepository;
+import de.tum.in.www1.artemis.service.dto.AbstractBuildResultNotificationDTO;
+
+public abstract class AbstractContinuousIntegrationService implements ContinuousIntegrationService {
+
+    private final Logger log = LoggerFactory.getLogger(AbstractContinuousIntegrationService.class);
+
+    protected ProgrammingSubmissionRepository programmingSubmissionRepository;
+
+    public AbstractContinuousIntegrationService(ProgrammingSubmissionRepository programmingSubmissionRepository) {
+        this.programmingSubmissionRepository = programmingSubmissionRepository;
+    }
+
+    /**
+     * Retrieves the submission that is assigned to the specified participation and its commit hash matches the one from
+     * the build result.
+     *
+     * @param participationId id of the participation
+     * @param buildResult     The build results
+     * @return The submission or empty no submissions exist
+     */
+    protected Optional<ProgrammingSubmission> getSubmissionForBuildResult(Long participationId, AbstractBuildResultNotificationDTO buildResult) {
+        var submissions = programmingSubmissionRepository.findAllByParticipationId(participationId);
+        if (submissions.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return submissions.stream().filter(theSubmission -> {
```

Comment:
Nice :D that fixes the issue we had with the duplicate submissions in the database.